refactor: align Feature2Mesh and Feature2Texture

zarov opened the pull request:
## Description

The two modules are very similar: they take a feature or an array of feature as arguments, and return either a mesh or a texture. It makes sense to regroup them under the same scheme.

The documentation was completed in the same time.

No breaking change as the old function has been marked as deprecated.

## Question

I'm not quite happy with the name of the common method (`convert`), and I wish to find a better name. In the same vein, I would like to take the opportunity to move those two module in `src/Core/newFolder` to start the gathering of all "stylizing/symbolizing" modules (as in #695 maybe). Any suggestion ?

zarov committed:
refactor: align Feature2Mesh and Feature2Texture

The two modules are very similar: they take a feature or an array of
feature as arguments, and return either a mesh or a texture. It makes
sense to regroup them under the same scheme.

They have been moved to another folder, src/Transform. They are not
available anymore under Feature2*.convert, but rather directly calling
feature2Mesh() or feature2Texture().

The documentation was completed in the same time.

BREAKING CHANGE: Feature2Mesh.convert doesn't exist anymore, in favor of
feature2Mesh, and the same applies for
Feature2Texture.createTextureFromFeature in favor of feature2Texture.

mbredif commented:
Thanks for this, that definitely goes in the right direction !
I agree '`convert`' might not be the best name ever. '`transform`' might be too generic as well, '`symbolize`' or '`stylize`' might be more explicit... If there is no agreement on a name, maybe we can just export the function rather than an object encapsulating a function.

#695 hints at a `src/Symbolizer` directory.

zarov commented:
> maybe we can just export the function rather than an object encapsulating a function.

As in `export default function`, and later call `Feature2Mesh/Texture()` ? Sounds good to me. I see no others methods that could be exported from this module, as it is quite long at the moment, so no problem to limit it to a function.

peppsac commented:
Suggestion: could you use this PR to add a test covering the features you're modifying? For instance add a test to `globe_vector.js` that would verify that picking the various rasterized elements (that use `Feature2Texture`) works as expected?

zarov commented:
@peppsac is d1805f3 adding a test as you expected ? I'll add more after.

zarov commented:
I pushed in the last commit the proposition of @mbredif, but the downside is we can't have a deprecation warning on `createTextureFromFeature`. WDYT ?

zarov commented:
@iTowns/reviewers anybody has an opinion here ?

peppsac commented on jsdoc-config.json:
```diff
@@ -23,6 +23,7 @@
 
             "src/Renderer/ColorLayersOrdering.js",
             "src/Renderer/ThreeExtended/Feature2Mesh.js",
+            "src/Renderer/ThreeExtended/Feature2Texture.js",
```
:+1:  :)

peppsac commented on src/Renderer/ThreeExtended/Feature2Mesh.js:
```diff
@@ -331,6 +331,8 @@ function geometryToExtrudedPolygon(geometry, properties, options, multiGeomAttri
  * @param {object|function} options.color - define per feature color
  * @return {THREE.Mesh} mesh
+ *
+ * @ignore
```
It's a bit weird to bother writing docs using jsdoc syntax and then add a `@ignore` tag.

IMHO, I'm fine with not publishing internal code documentation, but in this case I'd suggest to use a lighter (= easier to read by a human) doc format.
In this case it could be:
```js
// Convert a Feature geometry to a three.js mesh.
// Supported attributes (as constant or functions) in options are: altitude, extrude and color
```

zarov replied:
The goal was indeed to not publish this internal documentation, and not bloating the PR with a change here with `//`, but I'll add it if we're okay on it.

autra replied:
Why do you consider this as internal documentation? People can import every individual files in iTowns if they wish to. Imo this is completely public (and imo again, it's good that way).

zarov replied:
This part is kind of a duplicate of the function documentation below, so I don't see the point of exposing it.

peppsac replied:
> Why do you consider this as internal documentation

Becaue this function is not exported - so not accessible outside of Feature2Mesh.

autra replied:
Oh I see, sorry I misunderstood, you're actually documenting a private function. Is the current theme exposing these comments? I intent to hide them in the future theme.

peppsac commented on src/Renderer/ThreeExtended/Feature2Mesh.js:
```diff
@@ -415,28 +417,33 @@ function featureCollectionToThree(featureCollection, options) {
 
 /**
- * @module Feature2Mesh
+ * Return a function that converts [Features]{@link module:GeoJsonParser} to
```
Maybe we could ease the usage of this, and stop returning a function?

If a user needs to attach `options` we could instead support `Feature2Mesh.convert.bind(options)`?

zarov replied:
I'm all for it, I was afraid to break too much the current API.

peppsac reviewed (commented):
This PR is a good idea.

I left a couple suggestion regarding the API and the `@ignore` thing.

autra commented on examples/globe_wfs_extruded.html:
```diff
@@ -94,7 +94,7 @@
                 type: 'geometry',
                 update: itowns.FeatureProcessing.update,
-                convert: itowns.Feature2Mesh.convert({
+                convert: itowns.Feature2Mesh({
```
Little detail, but can we then change the name of the export to `feature2Mesh` without capitalization?

zarov replied:
Yep

autra commented:
> I would like to take the opportunity to move those two module in src/Core/newFolder to start the gathering of all "stylizing/symbolizing" modules (as in #695 maybe).

Good idea! I think you can drop the `Core` part, and put them directly into src/newFolder. We must find a balance between too many top level folders and too many folder level of course, but I think iTowns currently needs a flatter hierarchy (and I generally find them easier for newcomers, ex THREE.js has 15 folders under `src/`, and it' actually quite easy to find what you're looking for imho).

(a little detail, but I'd also like to drop folder capitalization `Core` -> `core` etc... I don't see the point of having them capitalized)

As for `newFolder`, I don't have any really great idea. Are these 2 files already stylizer according to you? Or will the styling process be after this transformation?

zarov commented:
> As for newFolder, I don't have any really great idea. Are these 2 files already stylizer according to you? Or will the styling process be after this transformation?

Currently I don't really see if the styling would be directly done inside those transformers, and I guess if we don't want to over bloat those files, it should be written outside. What about something like `src/transform` ?

zarov commented:
So in the last commit I did a few things:
- moved the `Feature2Mesh/Texture` files to `src/Transform/feature2Mesh/Texture`
- removed the option `convert` when adding a WFS layer (see the `wfs.html` for example), in favor of a `style` property of the layer, which is an object containing options for the styling of the layer (may be completed in the future when styling will be more advanced)
- the choice of using either the `feature2Mesh` or `feature2Texture` is done when adding the layer, depending on its type. The chosen method is set in `layer.transform`.
- `feature2Mesh/Texture` are now resolving directly a Promise, without getting the trouble to first return a function to execute
- `feature2Mesh` is not exposed in the public API anymore

As I feel there are a few breaking changes here, I would like at least two reviews before merging ! ping @iTowns/reviewers and @mbredif

autra commented:
> removed the option convert when adding a WFS layer (see the wfs.html for example), in favor of a style property of the layer, which is an object containing options for the styling of the layer (may be completed in the future when styling will be more advanced)

A couple of thoughts about this:

I still have some trouble drawing the lines between transformers and stylizers, I'm not even sure we should draw it.

Also, transformers could be chained. For instance, here we get geojson, we transform it once to get some features, then a second time to get a mesh, then a third time to extrude it, then a fourth time to set its altitude. Maybe this could be explicit?

But also we have colors and such, which looks like a stylizer, but the more I think about it, the more it looks like yet another transformation to me. 

So instead of "style", maybe in the future we could aim to have

```js
const transformation = chainTransformation(
        toFeature,
        toMesh,
        extrusion,
        altitude.bind(null, altitudeLine),
        color.bind(null, colorLine));
globeView.addLayer({
    name: 'lyon_tcl_bus',
    type: 'geometry',
    update: itowns.FeatureProcessing.update,
    transform: transfomation,
    linewidth: 5,
    filter: acceptFeatureBus,
    url: 'https://download.data.grandlyon.com/wfs/rdata?',
    protocol: 'wfs',
    version: '2.0.0',
    id: 'WFS Bus lines',
    typeName: 'tcl_sytral.tcllignebus',
    level: 9,
    projection: 'EPSG:3946',
    extent: {
        west: 1822174.60,
        east: 1868247.07,
        south: 5138876.75,
        north: 5205890.19,
    },
    format: 'geojson',
});

```
Where each transformation as an output format and a list of supported input format (and you could have adapters in case one transformer returns an array from a single input element). WDYT about this kind of pipeline API? 

Imo the main advantages:
- it's more explicit (like a lego, whereas a configuration is always "interpreted" somehow)
- it's more flexible (a config only allows you to do what is supported)
- it's more extensible (you could write your own transformer)
- it's also coherent with the `filter` property, and could also be coherent with a futur `animate` property (that'll serve to animate a layer)?

If you like it, then maybe Feature2Mesh could be seen as our first transformer, before getting splitted in the future.

In short, I think I preferred the previous version (nearer to what I propose here, with some naming to change).

Some questions remains:
- is it feasible?
- is it worthy? :-)

autra commented on src/Core/View.js:
```diff
@@ -180,6 +182,8 @@ function _preprocessLayer(view, layer, provider, parentLayer) {
         defineLayerProperty(layer, 'opacity', 1.0);
         defineLayerProperty(layer, 'sequence', 0);
+
+        layer.transform = feature2Texture;
```
layer.type color is also used for simple texture coming from WMTS, TMs and WMS, btw. It will still work with that, but it might be a bit confusing.

zarov replied:
Maybe then we could change the type if we want to use `feature2Texture`, something like `raster` (and then drop the `RasterProvider`)

autra replied:
It's a bit confusing because I feel like we're mixing input and output type. Regular color layers are `color` all the way, whereas in our case, the input type is `vector`, but the output type is `color`... 

Maybe just have:
```js

  if (layer.type == 'color') {
    layer.transform = feature2Texture;
  } else if (layer.type == 'geometry') {
    layer.transform = feature2Mesh;
  }
```  
in WFSProvider.preprocess ? 

After all, the input format is determined - or at least constrained - by the protocol, and it is only currently supported for WFS layer. 

Or another idea (that'll take a bit more work maybe), make the `preprocessing` phase marks each layer with the output format of the protocol. So for layers with protocol set to 'raster', 'wfs' (and '3dtiles' in the future?), that would be `layer.protocolOutputFormat = 'feature'` (ok maybe a better name needs to be found), and for WMS that would be `'texture'` for instance. Then in View.js:178, you'd have:

```js
if (layer.protocolOutputFormat  == 'feature') {
   // logic to set the right layer.transform according to the type
}
```

Also the advantage is that it'll be easy to have this property set by other providers like `FileProvider`, which would be one good step (maybe the final one actually) towards the removal of RasterProvider (parsers would return Feature instead of Mesh directly). WDYT?

peppsac commented:
FWIW I'd prefer to stick with the style vocabulary, for one main reason: it's closer to existing APIs.

zarov commented:
> Also, transformers could be chained. For instance, here we get geojson, we transform it once to get some features, then a second time to get a mesh, then a third time to extrude it, then a fourth time to set its altitude.

The geojson to features is the goal of the GeojsonParser, and as it is/will be specified by the `format`, it is already kinda chained. For the following things, the type `geometry` tells that we want a mesh, and we apply some options to it (`style`). It seems already chained to me.

Flexibility can be seducing, but it can quickly became complicated to implement and follow it, but also not efficient in performance. We already aim for an enough divided cutting in the API imho.

autra commented:
> It seems already chained to me.

Yes it is chained of course, but users don't have direct control over that chain.

> Flexibility can be seducing, but it can quickly became complicated to implement and follow it

Good point.

To better explain my motivation, I'm a bit disturbed by this `style` object which:
- seems specific to wfs layers in its current form.
- is not extensible by external app devs.

That being said, I don't have other concrete non supported use case yet (is there anything else to do with buildings than extrusion, altitude and colors?), and also, I don't see a type of data other than vectors that would need styling. So maybe I'm worrying for nothing :-) 

Nothing preventing this PR to go on, anyway.

autra commented on src/Transform/feature2Texture.js:
```diff
@@ -95,37 +102,55 @@ function drawFeature(ctx, feature, origin, scale, extent, style = {}) {
+ * @param {number} options.size - The size of the texture, in pixels. The
+ * resulting texture will be a square, and it is highly recommended to use a
+ * power of 2 for GPU optimization purpose.
+ * @param {Object|function} options.style - The style to apply to the
```
Oh I didn't realize functions were still supported. Which should actually gives all the possibilities I would ever want :-)

autra commented on src/Provider/RasterProvider.js:
```diff
@@ -31,7 +31,11 @@ function createTextureFromVector(tile, layer) {
         const coords = tile.extent;
         const result = { pitch: new THREE.Vector4(0, 0, 1, 1) };
-        result.texture = Feature2Texture.createTextureFromFeature(layer.feature, tile.extent, 256, layer.style);
+        result.texture = feature2Texture(layer.feature, {
```
so following the idea above, RasterProvider would set `layer.outputFormat = 'feature'` and use layer.transform here. And now I'm wondering if `layer.transform` shouldn't be called in the Scheduler, on line 54:

```js
return provider.executeCommand(cmd).then((result) => {
    if (cmd.layer.transform) {
        cmd.layer.transform(result);
    }
    this.counters.executing--;
    cmd.resolve(result);
    // only count successul commands
    this.counters.executed++;
}, (err) => {
    this.counters.executing--;
    cmd.reject(err);
    this.counters.failed++;
    if (__DEBUG__ && this.counters.failed < 3) {
        console.error(err);
    }
});
```

zarov replied:
> layer.transform shouldn't be called in the Scheduler

It should, but not in this PR. We can have the same principle as the current providers and the future parsers.

autra commented on src/Transform/feature2Mesh.js:
```diff
@@ -415,28 +407,31 @@ function featureCollectionToThree(featureCollection, options) {
-        };
-    },
-};
+export default function (feature, options = {}) {
```
Let's rename `options` to `style` then.

autra replied:
(and maybe also support functions, though maybe not in this PR)

zarov replied:
Yep, that would be better.

zarov replied:
In fact I prefer to let the `options` naming. It makes sense to change it in `feature2Mesh`, but not in `feature2Texture`. I feel more that I should change the `style` options when adding a layer. Any suggestion on a name ?

autra replied:
Yes indeed.

We might need something more generic: `paint` like mapboxgl? Or `draw`? `drawOptions`?

@peppsac, you said you wanted to stay with `style` though?

peppsac replied:
I don't mind `options` here (but I prefer `style` when its declared within the layer)

autra reviewed (commented):
I should have started a review sooner :-)

overall LGTM, though I hope we'll find something for the `transform` mechanism.

Don't hesitate to tell if you find out my proposal implies the rewriting of half of itowns :-)

zarov commented:
> (is there anything else to do with buildings than extrusion, altitude and colors?)

Yes, texturing ! I hope to have a `texture` or even `material` as another option someday.

zarov commented:
Closing this PR, as this is too old to still work on.
